components/scream/src/physics/spa/spa_functions_impl.hpp

```diff
@@ -0,0 +1,168 @@
+    });
+    team.team_barrier();
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nk_pack), [&] (int k) {
+      const auto& aer_g_sw_beg_sub           = ekat::subview(data_beg.AER_G_SW, i);
+      const auto& aer_g_sw_end_sub           = ekat::subview(data_end.AER_G_SW, i);
+      const auto& aer_g_sw_src_sub           = ekat::subview(aer_g_sw_src, i);
+
+      const auto& aer_ssa_sw_beg_sub         = ekat::subview(data_beg.AER_SSA_SW, i);
+      const auto& aer_ssa_sw_end_sub         = ekat::subview(data_end.AER_SSA_SW, i);
+      const auto& aer_ssa_sw_src_sub         = ekat::subview(aer_ssa_sw_src, i);
+  
+      const auto& aer_tau_sw_beg_sub         = ekat::subview(data_beg.AER_TAU_SW, i);
+      const auto& aer_tau_sw_end_sub         = ekat::subview(data_end.AER_TAU_SW, i);
+      const auto& aer_tau_sw_src_sub         = ekat::subview(aer_tau_sw_src, i);
+
+      Kokkos::parallel_for(
+        Kokkos::ThreadVectorRange(team, nswbands), [&] (Int n) {
+        aer_g_sw_src_sub(n,k)   = aer_g_sw_beg_sub(n,k)   + slope * (aer_g_sw_end_sub(n,k)   - aer_g_sw_beg_sub(n,k));
+        aer_ssa_sw_src_sub(n,k) = aer_ssa_sw_beg_sub(n,k) + slope * (aer_ssa_sw_end_sub(n,k) - aer_ssa_sw_beg_sub(n,k));
+        aer_tau_sw_src_sub(n,k) = aer_tau_sw_beg_sub(n,k) + slope * (aer_tau_sw_end_sub(n,k) - aer_tau_sw_beg_sub(n,k));
+      });
+    });
+    team.team_barrier();
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nk_pack), [&] (int k) {
+      const auto& aer_tau_lw_beg_sub         = ekat::subview(data_beg.AER_TAU_LW, i);
+      const auto& aer_tau_lw_end_sub         = ekat::subview(data_end.AER_TAU_LW, i);
+      const auto& aer_tau_lw_src_sub         = ekat::subview(aer_tau_lw_src, i);
+
+      Kokkos::parallel_for(
+        Kokkos::ThreadVectorRange(team, nlwbands), [&] (Int n) {
+        aer_tau_lw_src_sub(n,k) = aer_tau_lw_beg_sub(n,k) + slope * (aer_tau_lw_end_sub(n,k) - aer_tau_lw_beg_sub(n,k));
+      });
+    });
+    team.team_barrier();
+    }
+  });
+  Kokkos::fence();
+
+  using LIV = ekat::LinInterp<Real,Spack::n>;
+  Real minthreshold = 0.0;  // Hard-code a minimum value for aerosol concentration to zero.
+
+  LIV VertInterp(ncols,pressure_state.nlevs,nlevs,minthreshold);
+  Kokkos::parallel_for("vertical-interp-spa",
+    VertInterp.m_policy,
+    KOKKOS_LAMBDA(typename LIV::MemberType const& team) {
+    const int i = team.league_rank();
+    VertInterp.setup(team,
+                     ekat::subview(p_src,i),
+                     ekat::subview(pressure_state.pmid,i));
+    team.team_barrier();
+    VertInterp.lin_interp(team,
+                          ekat::subview(p_src,i),
+                          ekat::subview(pressure_state.pmid,i),
+                          ekat::subview(ccn3_src,i),
+                          ekat::subview(data_out.CCN3,i));
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nlwbands), [&] (int n) {
+      VertInterp.lin_interp(team,
```

Comment:
I don't think you can call lin_interp from _within_ a TeamThreadRange loop. From what I see in ekat, lin interp dispatches _its own_ TeamThreadRange loop.

@jgfouca Do you think it makes sense (and is it possible?) to make lin_interp templated on the policy, so that one can call it with either a TTR (and vec_length=1) or from a TVR?

Reply:
@jgfouca , I'll wait for your comment on this before making any changes.

Reply:
@bartgol , @AaronDonahue , you're right, LinInterp expects to be able to dispatch its own TeamThreadRange kernel. I'm actually surprised that this code worked. I will look at changing the interface.

Reply:
Another thought, would it make sense to change these team thread range kernels to regular for-loops?

Reply:
Spoke with @bartgol offline and we agreed to pursue the LinInterp templating

Reply:
@jgfouca , thanks for getting back on this.  So, should we wait on the change in EKAT and then incorporate the updated EKAT in this PR?

Reply:
See https://github.com/E3SM-Project/EKAT/pull/124

Reply:
@AaronDonahue , yes, lets wait and use the new LinInterp API which better-supports your use case here. I actually don't think the code you have here will work on GPU, but I could be wrong.

Reply:
huh, it seems to pass the tests on Weaver, so I wonder if it's not actually running that section of code

Reply:
I don't think kokkos (or cuda) checks that you are nesting the || layers correctly. You can probably write a TTR inside a TVR inside a TTR, and kokkos/cuda would not complain (but you'd probably get garbage out of it).

Reply:
@AaronDonahue , now that you have the latest EKAT, you'll need a slight modification to all the lin_interp calls that happen within a TeamThreadRange.  I will add some comments to help you.